automation/orchestrator.go

```diff
@@ -0,0 +1,157 @@
+package automation
+
+import (
+	"context"
+	"fmt"
+
+	golog "github.com/ipfs/go-log"
+	"github.com/qri-io/ioes"
+	"github.com/qri-io/qri/automation/run"
+	"github.com/qri-io/qri/automation/workflow"
+	"github.com/qri-io/qri/event"
+	"github.com/qri-io/qri/profile"
+)
+
+var (
+	log = golog.Logger("automation")
+)
+
+// OrchestratorOptions encapsulate runtime configuration for NewOrchestrator
+type OrchestratorOptions struct {
+	WorkflowStore workflow.Store
+}
+
+// Save persists the dataset that results from executing a workflow transform
+type Save func(ctx context.Context, streams ioes.IOStreams, w *workflow.State, runID string) error
+
+// SaveFactory is a function that produces a Save function
+type SaveFactory func(ctx context.Context) Save
+
+// Apply executes an ephemeral workflow transform
+type Apply func(ctx context.Context, streams ioes.IOStreams, w *workflow.State) error
+
+// ApplyFactory is function that produces an Apply function
+type ApplyFactory func(ctx context.Context) Apply
+
+// Orchestrator manages automation in qri
+type Orchestrator struct {
+	workflows    workflow.Store
+	saveFactory  SaveFactory
+	applyFactory ApplyFactory
+	bus          event.Bus
+	cancel       context.CancelFunc
+}
+
+// Shutdown tears down the orchestrator system
+// must be called to ensure all processes have be closed correctly
+func (o *Orchestrator) Shutdown() {
+	// TODO (ramfox): when we have added a way to unsubscribe from a bus, this is where we should do it
+	o.cancel()
+}
+
+// handleTrigger calls `RunWorkflow` when an `event.ETTriggerWorkflow` event is fired
+func (o *Orchestrator) handleTrigger(ctx context.Context, e event.Event) error {
+	if e.Type == event.ETTriggerWorkflow {
+		wid, ok := e.Payload.(workflow.ID)
+		if !ok {
+			return fmt.Errorf("handleTrigger: expected event.Payload to be a workflow.ID: %v", e.Payload)
+		}
+		return o.RunWorkflow(ctx, wid)
+	}
+	return nil
+}
+
+// RunWorkflow runs the given workflow
+func (o *Orchestrator) RunWorkflow(ctx context.Context, wid workflow.ID) error {
+	log.Debugf("RunWorkflow, workflow ID: %s", wid)
+	save := o.saveFactory(ctx)
+	wf, err := o.GetWorkflow(wid)
+	if err != nil {
+		log.Debugf("RunWorkflow: error getting workflow from store: %s", err)
+		return err
```

Comment:
wrap errors to add context:

```suggestion
		log.Debugw("RunWorkflow: getting workflow from store", "err", err)
		return fmt.Errorf("getting stored workflow: %w", err)
```